sw/host/opentitanlib/src/transport/ti50emulator/mod.rs

```diff
@@ -0,0 +1,183 @@
+};
+
+mod emu;
+mod gpio;
+mod i2c;
+mod spi;
+mod uart;
+
+use crate::transport::ti50emulator::emu::Ti50SubProcess;
+
+pub struct Ti50Emulator {
+    inner: Rc<RefCell<Inner>>,
+}
+
+impl Ti50Emulator {
+    /// Create new instance of [`Ti50Emulator`] based on provided parameters.
+    pub fn open(
+        executable_directory: PathBuf,
+        executable: &String,
+        instance_prefix: &String,
+    ) -> anyhow::Result<Self> {
+        let tstamp = SystemTime::now().duration_since(UNIX_EPOCH)?;
+        let instance_name = format!(
+            "{}_{}_{}_{}",
+            instance_prefix,
+            process::id(),
+            tstamp.as_secs(),
+            tstamp.as_nanos()
+        );
+
+        let mut instance_directory = PathBuf::from("/tmp");
+        instance_directory.push(&instance_name);
+        let resource_directory = instance_directory.join("resources");
+        let runtime_directory = instance_directory.join("runtime");
+
+        info!("Initializing Ti50Emulator instance:{}", instance_name);
+        fs::create_dir(&instance_directory)?;
+        fs::create_dir(&resource_directory)?;
+        fs::create_dir(&runtime_directory)?;
+
+        Ok(Self {
+            inner: Rc::new(RefCell::new(Inner {
+                instance_directory: instance_directory.clone(),
+                _resource_directory: resource_directory,
+                _runtime_directory: runtime_directory,
+                _executable_directory: executable_directory,
+                _executable: Some(executable.clone()),
+                emulator: None,
+                spi_map: HashMap::new(),
+                gpio_map: HashMap::new(),
+                i2c_map: HashMap::new(),
+                uart_map: HashMap::new(),
+            })),
+        })
+    }
+}
+
+impl Drop for Ti50Emulator {
+    fn drop(&mut self) {
+        info!(
+            "Clenup Ti50Emulator instance directory: {}",
+            self.inner.borrow().instance_directory.display()
+        );
+        if let Err(e) = fs::remove_dir_all(&self.inner.borrow().instance_directory) {
+            error!("Can't remove instance directory error: {}", e)
+        }
+    }
+}
+
+/// Structure representing internal state of emulator
+struct Inner {
+    /// Path of parent directory representing instance.
+    instance_directory: PathBuf,
+    /// Path to a directory holding unmodified files (for example flash image).
+    _resource_directory: PathBuf,
+    /// Path to a directory containing working copies of image files and socket
+    /// used by the sub-process.
+    _runtime_directory: PathBuf,
+    /// Path to directory holding executable files - kernel, applications.
+    _executable_directory: PathBuf,
+    /// Current name of kernel executable files.
+    _executable: Option<String>,
+    /// `Emualtor` instance
+    emulator: Option<Rc<dyn Emulator>>,
+    /// HashMap for holding all instance of SPI indexed by its symbolic name.
```

Comment:
Can out double-check your grammar? I think you meant "Mapping of SPI handles by their symbolic names."

Reply:
ok.